### ISSUE TITLE
Canonical way to run inference on a trained model?

### ISSUE THREAD

**Issue by taketwo** [dnouri/skorch#238]
Let's say I have a trained `NeuralNet` model and one sample of data, and I want to infer network's output. Ideally, I'd like to have a method that:

1. Accepts non-batched data.
2. Handles cpu/gpu transfers transparently
2a. Data are automatically transferred to gpu if network is on gpu
2b. Output is on the same device as input (or possible to select explicitly)
3. In case the network has several output values, gives me all of them.
4. Does not compute gradients.

There are several methods with names that sound suitable for the task (`infer`, `forward`, `predict`), however none of them fits all the requirements I outlined. Despite having the least "suitable" name, `forward` actually satisfies most of the requirements, though I still need to explicitly unsqueeze/squeeze my data to use it.

Is there anything I missed?

**Comment by BenjaminBossan** [comment]
Good question, I will think about it the next time I've got some time. Meanwhile, could you maybe say what you want to do exactly with the output. To me, it sounds like maybe you would like to run a web app that serves model predictions. Is that so?

**Comment by taketwo** [comment]
> it sounds like maybe you would like to run a web app that serves model predictions

This is certainly one of the potential use cases. However, my situation is different. I am debugging a network that processes 3D point clouds. To understand what's going on and whether it learns anything useful, in the end of each epoch I try to visualize the output and some intermediate representations (which for convenience I made a part of the output). These are computed on a single test example.

**Comment by benjamin-work** [comment]
After considering your case, I would guess that `forward_iter` would be the best fit. It handles the boilerplate, allows you to choose the device and whether you want gradients (via `training`), and, if you have multiple outputs, it will return all of them. The method returns a generator, so you will need to materialize the output, though.

What this method does not do is handle single sample data. I am, however, not sure whether we need a dedicated method for that. If I was in your shoes, I would either try to slice the data while keeping the dimension (`X[n:n + 1]`) or write a custom function that adds and removes the batch dimension. It is probably not trivial from the skorch side to accept data that drops the batch dimension, since the module needs it, so skorch would need to re-add it, which is non-trivial for some data types.

One thing you could also consider: If your data handling is very easy, you could call the module directly:

```
x = torch.tensor(X[n:n + 1]).to(device)
with torch.no_grad():
    output, intermediate = net.module_(x)
```

PS: If you're interested in web services, keep an eye on #237.

**Comment by taketwo** [comment]
Thanks for the suggestions. Are there any reasons to prefer `forward_iter` over `forward`? The latter takes care of materializing the output, so even less boilerplate for me to write.

> It is probably not trivial from the skorch side to accept data that drops the batch dimension, since the module needs it, so skorch would need to re-add it, which is non-trivial for some data types.

Agree, also the batch dimension would need to be stripped off from the output(s).

**Comment by taketwo** [comment]
I see, thanks.

If you think it might be helpful, I can add a note to `predict`/`predict_proba` docstrings suggesting to use `forward` in the case of multiple outputs. Otherwise I'd just close this.

**Comment by benjamin-work** [comment]
> I can add a note to predict/predict_proba docstrings suggesting to use forward in the case of multiple outputs

That is a good idea.
